packages/native/src/components/WebView.tsx

```diff
@@ -32,32 +36,115 @@ const NativeWebView: React.FC<WebViewProps> = ({
 
+  const [cameraPermissions, setCameraPermissions] =
+    useState<null | PermissionResponse>(null);
+
+  const [microphonePermissions, setMicrophonePermissions] =
+    useState<null | PermissionResponse>(null);
+
   const videoChatProps = optimizeVideoChat
     ? {
         allowsInlineMediaPlayback: true,
         domStorageEnabled: true,
         javaScriptEnabled: true,
+        mediaCapturePermissionGrantType: "grant", // so iOS uses system settings
         mediaPlaybackRequiresUserAction: false,
         startInLoadingState: true,
       }
-    : ({} as Record<string, boolean>);
+    : ({} as Record<string, boolean | string>);
 
   const onMessage = (event: WebViewMessageEvent) =>
     setHeight(Number(event.nativeEvent.data));
 
+  const getAndSetPermissions = async (
+    currentState: null | PermissionResponse,
+    setCurrentState: Dispatch<SetStateAction<null | PermissionResponse>>,
+    getPermission: () => Promise<PermissionResponse>,
+    requestPermission: () => Promise<PermissionResponse>
+  ) => {
+    const currentPermission = currentState ?? (await getPermission());
+
+    if (currentPermission.granted || !currentPermission.canAskAgain) {
+      setCurrentState(currentPermission);
+    } else {
+      setCurrentState(await requestPermission());
+    }
+  };
+
+  const getAndSetCameraAndMicrophonePermissions = async () => {
+    await getAndSetPermissions(
+      cameraPermissions,
+      setCameraPermissions,
+      Camera.getCameraPermissionsAsync,
+      Camera.requestCameraPermissionsAsync
+    );
+
+    await getAndSetPermissions(
+      microphonePermissions,
+      setMicrophonePermissions,
+      Camera.getMicrophonePermissionsAsync,
+      Camera.requestMicrophonePermissionsAsync
+    );
+  };
+
+  const selectComponent = () => {
+    if (
+      !optimizeVideoChat ||
+      (cameraPermissions?.granted && microphonePermissions?.granted)
+    ) {
+      return (
+        <WebView
+          source={source}
+          style={[{ width: optimizeVideoChat ? width : undefined }, style]}
+          injectedJavaScript={injectFirst}
+          onMessage={onMessage}
+          {...videoChatProps}
+        />
+      );
+    } else if (
+      (!cameraPermissions?.granted && cameraPermissions?.canAskAgain) ||
+      (!microphonePermissions?.granted && microphonePermissions?.canAskAgain)
+    ) {
+      return (
+        <Button
+          title={"Press to enable Audio and/or Video permissions"}
+          onPress={getAndSetCameraAndMicrophonePermissions}
+        />
+      );
+    } else if (
+      (cameraPermissions?.status === "denied" &&
+        cameraPermissions?.canAskAgain === false) ||
+      (microphonePermissions?.status === "denied" &&
+        microphonePermissions?.canAskAgain === false)
+    ) {
+      return (
+        <Text>
+          {"Set the missing Audio and/or Video permissions in System Settings"}
+        </Text>
+      );
+    } else {
+      return <ActivityIndicator />;
+    }
+  };
```

Comment:
(nit) It's a small thing, but this function might be a bit more readable like this

```suggestion
  const selectComponent = () => {
    if (
      !optimizeVideoChat ||
      (cameraPermissions?.granted && microphonePermissions?.granted)
    ) {
      return (
        <WebView
          source={source}
          style={[{ width: optimizeVideoChat ? width : undefined }, style]}
          injectedJavaScript={injectFirst}
          onMessage={onMessage}
          {...videoChatProps}
        />
      );
    }
    
    if (
      (!cameraPermissions?.granted && cameraPermissions?.canAskAgain) ||
      (!microphonePermissions?.granted && microphonePermissions?.canAskAgain)
    ) {
      return (
        <Button
          title={"Press to enable Audio and/or Video permissions"}
          onPress={getAndSetCameraAndMicrophonePermissions}
        />
      );
    }
    
    if (
      (cameraPermissions?.status === "denied" &&
        cameraPermissions?.canAskAgain === false) ||
      (microphonePermissions?.status === "denied" &&
        microphonePermissions?.canAskAgain === false)
    ) {
      return (
        <Text>
          {"Set the missing Audio and/or Video permissions in System Settings"}
        </Text>
      );
    }
    
    return <ActivityIndicator />;
  };
```